### PR TITLE
[Attributor] Fix an issue that an access is skipped by mistake

### PR DESCRIPTION
**Pull request by shiltian**
When we check if an access can be skipped, there is a case that an
inter-procedural interference access exists after a dominant write. Currently we
rely on `AAInterFnReachability` to tell if the access can be reachable. If it is
not, we can safely skip the access. However, it is based on an assumption that
the AA exists. It is possible that the AA doesn't exist. In this case, we can't
safely assume the acess can be skipped because we have to assume the access can
reach. This can happen when `AAInterFnReachability` is not in the allowed AA
list when creating the attributor, such as AMDGPUAttributor.

### CODE REVIEW

**Comment by shiltian** [comment]
* **#101862** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/101862?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @shiltian and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by arsenm — COMMENTED** [review]
Testcase?

**Comment by shiltian** [comment]
> Testcase?

Well, the thing is, this can only be exposed in an Attributor use that doesn't have `AAInterFnReachability`. Regular `AttributorPass` will not have this issue. Only `AMDGPUAttributorPass` can do this, but the enabling of `AAIndirectCallInfo ` has been reverted so even that can't expose it.

**Comment by shiltian** [comment]
ping

**Comment by arsenm** [comment]
> > Testcase?
> 
> Well, the thing is, this can only be exposed in an Attributor use that doesn't have `AAInterFnReachability`. Regular `AttributorPass` will not have this issue. Only `AMDGPUAttributorPass` can do this, but the enabling of `AAIndirectCallInfo ` has been reverted so even that can't expose it.
> 
> What I can do I guess is still to include the test case in this PR and the enabling of `AAIndirectCallInfo ` in `AMDGPUAttributor` should not break it. Is that okay?

Pre-commit the test that would break in the next patch in its working state?

**Comment by shiltian** [comment]
> > > Testcase?
> > 
> > 
> > Well, the thing is, this can only be exposed in an Attributor use that doesn't have `AAInterFnReachability`. Regular `AttributorPass` will not have this issue. Only `AMDGPUAttributorPass` can do this, but the enabling of `AAIndirectCallInfo ` has been reverted so even that can't expose it.
> > What I can do I guess is still to include the test case in this PR and the enabling of `AAIndirectCallInfo ` in `AMDGPUAttributor` should not break it. Is that okay?
> 
> Pre-commit the test that would break in the next patch in its working state?

Hmm, I didn't follow.

**Comment by rovka** [comment]
> > Testcase?
> 
> Well, the thing is, this can only be exposed in an Attributor use that doesn't have `AAInterFnReachability`. Regular `AttributorPass` will not have this issue. Only `AMDGPUAttributorPass` can do this, but the enabling of `AAIndirectCallInfo ` has been reverted so even that can't expose it.

I'm not very familiar with these APIs, but is this something that would be easier to test with a C++ unit test? I see there are already some [attributor unit tests](https://github.com/llvm/llvm-project/blob/main/llvm/unittests/Transforms/IPO/AttributorTest.cpp).

**Comment by shiltian** [comment]
> > > Testcase?
> > 
> > 
> > Well, the thing is, this can only be exposed in an Attributor use that doesn't have `AAInterFnReachability`. Regular `AttributorPass` will not have this issue. Only `AMDGPUAttributorPass` can do this, but the enabling of `AAIndirectCallInfo ` has been reverted so even that can't expose it.
> 
> I'm not very familiar with these APIs, but is this something that would be easier to test with a C++ unit test? I see there are already some [attributor unit tests](https://github.com/llvm/llvm-project/blob/main/llvm/unittests/Transforms/IPO/AttributorTest.cpp).

Yes, that's a good point. However, this issue requires a target where a stack variable is not accessible by other threads (effectively a GPU). I can't add target triple in the unit test because it will break a build if the target is not enabled.

File: llvm/lib/Transforms/IPO/AttributorAttributes.cpp [hunk 1]
```diff
@@ -1331,7 +1331,7 @@ struct AAPointerInfoImpl
         // itself either.
         bool Inserted = ExclusionSet.insert(&I).second;
 
-        if (!FnReachabilityAA ||
```
Comment (jdoerfert) [inline llvm/lib/Transforms/IPO/AttributorAttributes.cpp:1334]: We don't need to insert/remove if the Reachability AA is not present. Simply skip the entire thing.

**Review by jdoerfert — APPROVED** [review]

**Comment by shiltian** [comment]
Hmm, I can't reproduce this issue on gfx1030.

**Comment by jdoerfert** [comment]
> Hmm, I can't reproduce this issue on gfx1030 as well as gfx90a.

Was a fluke: https://lab.llvm.org/buildbot/#/builders/73/builds/3244

**Pull request merged**